### ISSUE TITLE
Missing quantize gradient?

### ISSUE THREAD

**Issue by ranery** [zzzxxxttt/pytorch_DoReFaNet#6]
Hi, do you consider gradient quantization in your implementation?

**Comment by zzzxxxttt** [comment]
No, The gradient quantization is not included in my implementation. I think it is quite difficult to simulate quantized backpropagation using pytorch. Although it is possible (maybe using backward hook ?), the training speed will be extremely slow.

**Comment by ranery** [comment]
I rewrite your quantization function following ![official implementation](https://github.com/tensorpack/tensorpack/tree/master/examples/DoReFa-Net), the speed is acceptable:
````
def uniform_quantize(k, k_g):
  class qfn(torch.autograd.Function):

    @staticmethod
    def forward(ctx, input):
      if k == 32:
        out = input
      elif k == 1:
        out = torch.sign(input)
      else:
        n = float(2 ** k - 1)
        out = torch.round(input * n) / n
      return out

    @staticmethod
    def backward(ctx, grad_output):
      if k_g == 32:
        return grad_output
      else:
        x = grad_output.clone()
        rank = len(x.shape)
        assert rank is not None
        maxx = torch.max(torch.abs(x), 1, keepdim=True)
        for i in range(2, rank):
            maxx = torch.max(torch.abs(maxx[0]), i, keepdim=True)
        x = x / maxx[0]
        n = float(2**k_g - 1)
        x = x * 0.5 + 0.5 + torch.FloatTensor(x.shape).uniform_(-0.5 / n, 0.5 / n).cuda()
        x = torch.clamp(x, 0.0, 1.0)
        x = torch.round(x * n) / n - 0.5
        return x * maxx[0] * 2

  return qfn().apply
````

**Comment by zzzxxxttt** [comment]
How do you apply this quantization function? I think the gradients of all the intermediate tensors during forward propagation need to be quantized, not only the gradient of weights.

**Comment by zzzxxxttt** [comment]
ok, I mistakenly understand the gradient quantization, I just realized that only the gradient of weights needs to be quantized...

**Comment by HaFred** [comment]
> ok, I mistakenly understand the gradient quantization, I just realized that only the gradient of weights needs to be quantized...

Hi @zzzxxxttt @ranery , actually in the dorefa-net paper, the algorithm.1 shows that for grad quantization, only the grads of activations (not weights) are quantized in step. 12 by the STE. So the codes above might not be appropriate?
